cli/deployments/tekton.go

```diff
@@ -0,0 +1,260 @@
+	out, err = helpers.SpinnerWaitCommand(message,
+		func() (string, error) {
+			return helpers.KubectlApplyEmbeddedYaml(tektonTriggersYamlPath)
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("%s failed:\n%s", message, out))
+	}
+
+	message = "Install the tekton dashboard"
+	out, err = helpers.SpinnerWaitCommand(message,
+		func() (string, error) {
+			return helpers.KubectlApplyEmbeddedYaml(tektonDashboardYamlPath)
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("%s failed:\n%s", message, out))
+	}
+
+	message = "Waiting for registry certificates to be created in the eirini-workloads namespace"
+	out, err = helpers.SpinnerWaitCommand(message,
+		func() (string, error) {
+			out1, err := helpers.ExecToSuccessWithTimeout(
+				func() (string, error) {
+					return helpers.Kubectl("get secret -n eirini-workloads registry-tls-self-ca")
+				}, 300*time.Second, 3*time.Second)
+			if err != nil {
+				return out1, err
+			}
+
+			out2, err := helpers.ExecToSuccessWithTimeout(
+				func() (string, error) {
+					return helpers.Kubectl("get secret -n eirini-workloads registry-tls-self")
+				}, 300*time.Second, 3*time.Second)
+
+			return fmt.Sprintf("%s\n%s", out1, out2), err
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("%s failed:\n%s", message, out))
+	}
+
+	message = "Applying tekton staging resources"
+	out, err = helpers.SpinnerWaitCommand(message,
+		func() (string, error) {
+			return applyTektonStaging(c)
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("%s failed:\n%s", message, out))
+	}
+
+	emoji.Println(":heavy_check_mark: Tekton deployed")
+
+	return nil
+}
+
+func (k Tekton) GetVersion() string {
+	return fmt.Sprintf("pipelines: %s, triggers %s, dashboard: %s",
+		tektonPipelineReleaseYamlPath, tektonTriggersReleaseYamlPath, tektonDashboardYamlPath)
+}
+
+func (k Tekton) Deploy(c kubernetes.Cluster, options kubernetes.InstallationOptions) error {
+
+	_, err := c.Kubectl.CoreV1().Namespaces().Get(
+		context.Background(),
+		tektonDeploymentID,
+		metav1.GetOptions{},
+	)
+	if err == nil {
+		return errors.New("Namespace " + tektonDeploymentID + " present already")
+	}
+
+	emoji.Println(":ship:Deploying Tekton")
+	err = k.apply(c, options, false)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Tekton) Upgrade(c kubernetes.Cluster, options kubernetes.InstallationOptions) error {
+	_, err := c.Kubectl.CoreV1().Namespaces().Get(
+		context.Background(),
+		tektonDeploymentID,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		return errors.New("Namespace " + tektonDeploymentID + " not present")
+	}
+
+	emoji.Println(":ship:Upgrade Tekton")
+	return k.apply(c, options, true)
+}
+
+// The equivalent of:
+// kubectl get secret -n eirini-workloads registry-tls-self -o json | jq -r '.["data"]["ca"]' | base64 -d | openssl x509 -hash -noout
+// written in golang.
+func getRegistryCAHash(c kubernetes.Cluster) (string, error) {
```

Comment:
These methods are not properly "scoped" I think. The are available all over the `deployments` package although they are only relevant to this deployment. Maybe they should have this deployment as a receiver.